File: backend/app/schemas/race.py
"""
Race response schemas.

These schemas match the frontend race contract and use ``CamelModel`` so
their snake_case Python fields serialize to the expected camelCase names.
"""

from typing import List, Optional

from app.schemas.base import CamelModel


class RaceSummary(CamelModel):
    race_id: int
    year: int
    round: int
    name: str
    circuit_name: str
    country: Optional[str]
    date: str


class RaceResultRow(CamelModel):
    position: Optional[int]
    position_text: str
    driver_name: str
    constructor_name: str
    constructor_ref: str
    grid: int
    points: float
    laps: int
    status: str
    fastest_lap_time: Optional[str]


class QualifyingRow(CamelModel):
    position: Optional[int]
    driver_name: str
    constructor_name: str
    constructor_ref: str
    q1: Optional[str]
    q2: Optional[str]
    q3: Optional[str]


class PitStopRow(CamelModel):
    driver_name: str
    stop: int
    lap: int
    time: Optional[str]
    duration: Optional[str]


class LapTimePoint(CamelModel):
    lap: int
    driver_code: str
    seconds: float


class RaceDetail(CamelModel):
    race: RaceSummary
    results: List[RaceResultRow]
    qualifying: List[QualifyingRow]
    pit_stops: List[PitStopRow]
    lap_times: List[LapTimePoint]

File: frontend/src/types/race.ts
export interface RaceSummary {
  raceId: number;
  year: number;
  round: number;
  name: string;
  circuitName: string;
  country: string | null;
  date: string; // ISO date
  winnerDriverName?: string | null;
  winnerConstructorName?: string | null;
  winnerConstructorRef?: string | null;
  winnerConstructorAbbreviation?: string | null;
}

export interface RaceResultRow {
  position: number | null;
  positionText: string;
  driverName: string;
  constructorName: string;
  constructorRef: string;
  grid: number;
  points: number;
  laps: number;
  status: string;
  fastestLapTime: string | null;
}

export interface QualifyingRow {
  position: number | null;
  driverName: string;
  constructorName: string;
  constructorRef: string;
  q1: string | null;
  q2: string | null;
  q3: string | null;
}

export interface PitStopRow {
  driverName: string;
  stop: number;
  lap: number;
  time: string | null;
  duration: string | null;
}

export interface LapTimePoint {
  lap: number;
  driverCode: string;
  /** Lap time converted to seconds for charting (e.g. "1:23.456" -> 83.456). */
  seconds: number;
}

export interface RaceDetail {
  race: RaceSummary;
  results: RaceResultRow[];
  qualifying: QualifyingRow[];
  pitStops: PitStopRow[];
  lapTimes: LapTimePoint[];
}
